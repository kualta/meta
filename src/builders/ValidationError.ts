import type { z } from 'zod';

import { formatZodError } from '../formatters';

/**
 * An error that occurs when an object does not match the expected shape.
 */
export class ValidationError extends Error {
  name = 'ValidationError' as const;
}

/**
 * @internal
 */
export function evaluate<Input, Output>(result: z.SafeParseReturnType<Input, Output>): Output {
  if (result.success) {
    return result.data;
  }
  throw new ValidationError(formatZodError(result.error));
}
