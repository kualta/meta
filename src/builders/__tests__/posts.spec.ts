import { faker } from '@faker-js/faker';
import { describe, expect, it } from 'vitest';

import {
  MediaAudioMimeType,
  MediaImageMimeType,
  MediaVideoMimeType,
  MetadataLicenseType,
  MetadataTransactionType,
  ThreeDFormat,
} from '../../post/index.js';
import { geoUri } from '../../primitives.js';
import { ValidationError } from '../ValidationError.js';
import {
  article,
  audio,
  checkingIn,
  embed,
  event,
  image,
  link,
  liveStream,
  mint,
  shortVideo,
  space,
  story,
  textOnly,
  threeD,
  transaction,
  video,
} from '../posts.js';

describe('Given the post metadata builders', () => {
  describe('when setting tags on any post metadata', () => {
    it('should ensure they at 20 at most', () => {
      expect(() =>
        article({
          title: 'Great Question',
          content: 'What is the answer to life, the universe and everything?',
          tags: faker.word.words(21).split(''),
        }),
      ).toThrowError(ValidationError);
    });

    it('should ensure they are unique', () => {
      expect(() =>
        article({
          title: 'Great Question',
          content: 'What is the answer to life, the universe and everything?',
          tags: ['question', '42', '42'],
        }),
      ).toThrowError(ValidationError);
    });
  });

  describe(`when using the "${article.name}" builder`, () => {
    it('should return a valid TextOnlyMetadata', () => {
      const metadata = article({
        title: 'Great Question',
        content: `
        ## Heading

        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec a diam lectus. Sed sit amet ipsum mauris.

        ## Question

        What is the answer to life, the universe and everything?

        ## Answer

        42

        ![The answer](https://example.com/answer.png)
        `,
        tags: ['question', '42'],
        attachments: [
          {
            item: 'https://example.com/answer',
            type: MediaImageMimeType.PNG,
            altTag: 'The answer',
          },
        ],
      });

      expect(metadata).toMatchSnapshot({
        lens: {
          id: expect.any(String),
        },
      });
    });

    it(`should throw a "${ValidationError.name}" in case of invalid input`, () => {
      expect(() => article({ content: '' })).toThrowError(ValidationError);
    });
  });

  describe(`when using the "${audio.name}" builder`, () => {
    it('should return a valid AudioMetadata', () => {
      const metadata = audio({
        title: 'Great song!',
        audio: {
          item: 'https://example.com/song.mp3',
          type: MediaAudioMimeType.MP3,
          artist: 'John Doe',
          cover: 'https://example.com/cover.png',
        },
        attachments: [
          {
            item: 'https://example.com/song-1.mp3',
            type: MediaAudioMimeType.MP3,
            duration: 234,
          },
          {
            item: 'https://example.com/song-2.mp3',
            type: MediaAudioMimeType.MP3,
            duration: 345,
          },
        ],
      });

      expect(metadata).toMatchSnapshot({
        lens: {
          id: expect.any(String),
        },
      });
    });
  });

  describe(`when using the "${checkingIn.name}" builder`, () => {
    it('should return a valid CheckingInMetadata for Geo URI location', () => {
      const metadata = checkingIn({
        location: 'The Moon',
        position: geoUri({
          lat: 40.6892,
          lng: -74.0444,
        }),
      });

      expect(metadata).toMatchSnapshot({
        lens: {
          id: expect.any(String),
        },
      });
    });

    it('should return a valid CheckingInMetadata for PhysicalAddress', () => {
      const metadata = checkingIn({
        location: 'The Moon',
        address: {
          streetAddress: '1st Street',
          locality: 'New York',
          region: 'NY',
          postalCode: '10001',
          country: 'US',
        },
      });

      expect(metadata).toMatchSnapshot({
        lens: {
          id: expect.any(String),
        },
      });
    });
  });

  describe(`when using the "${embed.name}" builder`, () => {
    it('should return a valid EmbedMetadata', () => {
      const metadata = embed({
        embed: 'https://example.com/embed.html',
      });

      expect(metadata).toMatchSnapshot({
        lens: {
          id: expect.any(String),
        },
      });
    });
  });

  describe(`when using the "${event.name}" builder`, () => {
    it('should return a valid EventMetadata for Geo URI location', () => {
      const metadata = event({
        title: 'Where should we go next?',
        location: 'The Moon',
        position: geoUri({
          lat: 40.6892,
          lng: -74.0444,
        }),
        startsAt: '2028-10-01T00:00:00Z',
        endsAt: '2028-10-01T01:00:00Z',
        links: ['https://example.com/tickets.html'],
      });

      expect(metadata).toMatchSnapshot({
        lens: {
          id: expect.any(String),
        },
      });
    });

    it('should return a valid EventMetadata for PhysicalAddress', () => {
      const metadata = event({
        location: 'The Moon',
        address: {
          streetAddress: '1st Street',
          locality: 'New York',
          region: 'NY',
          postalCode: '10001',
          country: 'US',
        },
        startsAt: '2028-10-01T00:00:00Z',
        endsAt: '2028-10-01T01:00:00Z',
      });

      expect(metadata).toMatchSnapshot({
        lens: {
          id: expect.any(String),
        },
      });
    });
  });

  describe(`when using the "${image.name}" builder`, () => {
    it('should return a valid ImageMetadata', () => {
      const metadata = image({
        title: 'Touch grass',
        image: {
          item: 'https://example.com/image.png',
          type: MediaImageMimeType.PNG,
          altTag: 'Me touching grass',
          license: MetadataLicenseType.CCO,
        },
        attachments: [
          {
            item: 'https://example.com/image-1.png',
            type: MediaImageMimeType.PNG,
            license: MetadataLicenseType.CC_BY_NC,
            altTag: 'Me touching a tree',
          },
          {
            item: 'https://example.com/image-2.png',
            type: MediaImageMimeType.PNG,
            license: MetadataLicenseType.CC_BY_NC,
            altTag: 'The tree touching me',
          },
        ],
      });

      expect(metadata).toMatchSnapshot({
        lens: {
          id: expect.any(String),
        },
      });
    });
  });

  describe(`when using the "${link.name}" builder`, () => {
    it('should return a valid LinkMetadata', () => {
      const metadata = link({
        sharingLink: 'https://example.com/embed.html',
        content: 'Check out this cool website!',
      });

      expect(metadata).toMatchSnapshot({
        lens: {
          id: expect.any(String),
        },
      });
    });
  });

  describe(`when using the "${liveStream.name}" builder`, () => {
    it('should return a valid LiveStreamMetadata', () => {
      const metadata = liveStream({
        title: 'Live stream #1',
        liveUrl: 'https://example.com/live.html',
        playbackUrl: 'https://example.com/playback.html',
        startsAt: '2028-10-01T00:00:00Z',
      });

      expect(metadata).toMatchSnapshot({
        lens: {
          id: expect.any(String),
        },
      });
    });

    describe(`when using the "${mint.name}" builder`, () => {
      it('should return a valid MintMetadata', () => {
        const metadata = mint({
          content: 'Check out this NFT!',
          mintLink:
            'https://opensea.io/assets/0x495f947276749ce646f68ac8c248420045cb7b5e/1234567890',
        });

        expect(metadata).toMatchSnapshot({
          lens: {
            id: expect.any(String),
          },
        });
      });
    });
  });

  describe(`when using the "${space.name}" builder`, () => {
    it('should return a valid SpaceMetadata', () => {
      const metadata = space({
        title: 'Space #1',
        link: 'https://example.com/space.html',
        startsAt: '2028-10-01T00:00:00Z',
      });

      expect(metadata).toMatchSnapshot({
        lens: {
          id: expect.any(String),
        },
      });
    });
  });

  describe(`when using the "${story.name}" builder`, () => {
    it('should return a valid StoryMetadata', () => {
      const metadata = story({
        asset: {
          item: 'https://example.com/story.mp4',
          type: MediaVideoMimeType.MP4,
          cover: 'https://example.com/thumbnail.png',
          duration: 123,
          altTag: 'The story of my life',
        },
      });

      expect(metadata).toMatchSnapshot({
        lens: {
          id: expect.any(String),
        },
      });
    });
  });

  describe(`when using the "${threeD.name}" builder`, () => {
    it('should return a valid ThreeDMetadata', () => {
      const metadata = threeD({
        content: 'Check out this 3D model!',
        assets: [
          {
            format: ThreeDFormat.VRM,
            playerUrl: 'https://example.com/player.html',
            uri: 'https://example.com/model.zip',
            zipPath: 'foo/model.vrm',
          },
        ],
        tags: ['3d', 'vrm'],
      });

      expect(metadata).toMatchSnapshot({
        lens: {
          id: expect.any(String),
        },
      });
    });
  });

  describe(`when using the "${transaction.name}" builder`, () => {
    it('should return a valid TransactionMetadata', () => {
      const metadata = transaction({
        chainId: 1,
        txHash: '0x1234567890',
        content: 'Check out this transaction!',
        type: MetadataTransactionType.ERC20,
      });

      expect(metadata).toMatchSnapshot({
        lens: {
          id: expect.any(String),
        },
      });
    });
  });

  describe(`when using the "${textOnly.name}" builder`, () => {
    it('should return a valid TextOnlyMetadata', () => {
      const metadata = textOnly({ content: 'GM!' });

      expect(metadata).toMatchSnapshot({
        lens: {
          id: expect.any(String),
        },
      });
    });
  });

  describe(`when using the "${video.name}" builder`, () => {
    it('should return a valid VideoMetadata', () => {
      const metadata = video({
        title: 'Great video!',
        video: {
          item: 'https://example.com/video.mp4',
          type: MediaVideoMimeType.MP4,
          cover: 'https://example.com/thumbnail.png',
          duration: 123,
          altTag: 'The video of my life',
          license: MetadataLicenseType.CCO,
        },
        attachments: [
          {
            item: 'https://example.com/soundtrack.mp3',
            type: MediaAudioMimeType.MP3,
            license: MetadataLicenseType.CCO,
          },
        ],
      });

      expect(metadata).toMatchSnapshot({
        lens: {
          id: expect.any(String),
        },
      });
    });
  });

  describe(`when using the "${shortVideo.name}" builder`, () => {
    it('should return a valid VideoMetadata', () => {
      const metadata = shortVideo({
        title: 'Great video!',
        video: {
          item: 'https://example.com/video.mp4',
          type: MediaVideoMimeType.MP4,
          cover: 'https://example.com/thumbnail.png',
          duration: 123,
          altTag: 'The video of my life',
          license: MetadataLicenseType.CCO,
        },
      });

      expect(metadata).toMatchSnapshot({
        lens: {
          id: expect.any(String),
        },
      });
    });
  });
});
