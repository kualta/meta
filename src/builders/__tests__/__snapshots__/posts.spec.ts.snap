// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Given the post metadata builders > when using the "article" builder > should return a valid TextOnlyMetadata 1`] = `
{
  "$schema": "https://json-schemas.lens.dev/posts/article/3.0.0.json",
  "lens": {
    "attachments": [
      {
        "altTag": "The answer",
        "item": "https://example.com/answer",
        "type": "image/png",
      },
    ],
    "content": "## Heading

        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec a diam lectus. Sed sit amet ipsum mauris.

        ## Question

        What is the answer to life, the universe and everything?

        ## Answer

        42

        ![The answer](https://example.com/answer.png)",
    "id": Any<String>,
    "locale": "en",
    "mainContentFocus": "ARTICLE",
    "tags": [
      "question",
      "42",
    ],
    "title": "Great Question",
  },
}
`;

exports[`Given the post metadata builders > when using the "audio" builder > should return a valid AudioMetadata 1`] = `
{
  "$schema": "https://json-schemas.lens.dev/posts/audio/3.0.0.json",
  "lens": {
    "attachments": [
      {
        "duration": 234,
        "item": "https://example.com/song-1.mp3",
        "type": "audio/mpeg",
      },
      {
        "duration": 345,
        "item": "https://example.com/song-2.mp3",
        "type": "audio/mpeg",
      },
    ],
    "audio": {
      "artist": "John Doe",
      "cover": "https://example.com/cover.png",
      "item": "https://example.com/song.mp3",
      "type": "audio/mpeg",
    },
    "id": Any<String>,
    "locale": "en",
    "mainContentFocus": "AUDIO",
    "title": "Great song!",
  },
}
`;

exports[`Given the post metadata builders > when using the "checkingIn" builder > should return a valid CheckingInMetadata for Geo URI location 1`] = `
{
  "$schema": "https://json-schemas.lens.dev/posts/checking-in/3.0.0.json",
  "lens": {
    "id": Any<String>,
    "locale": "en",
    "location": "The Moon",
    "mainContentFocus": "CHECKING_IN",
    "position": "geo:40.6892,-74.0444",
  },
}
`;

exports[`Given the post metadata builders > when using the "checkingIn" builder > should return a valid CheckingInMetadata for PhysicalAddress 1`] = `
{
  "$schema": "https://json-schemas.lens.dev/posts/checking-in/3.0.0.json",
  "lens": {
    "address": {
      "country": "US",
      "locality": "New York",
      "postalCode": "10001",
      "region": "NY",
      "streetAddress": "1st Street",
    },
    "id": Any<String>,
    "locale": "en",
    "location": "The Moon",
    "mainContentFocus": "CHECKING_IN",
  },
}
`;

exports[`Given the post metadata builders > when using the "embed" builder > should return a valid EmbedMetadata 1`] = `
{
  "$schema": "https://json-schemas.lens.dev/posts/embed/3.0.0.json",
  "lens": {
    "embed": "https://example.com/embed.html",
    "id": Any<String>,
    "locale": "en",
    "mainContentFocus": "EMBED",
  },
}
`;

exports[`Given the post metadata builders > when using the "event" builder > should return a valid EventMetadata for Geo URI location 1`] = `
{
  "$schema": "https://json-schemas.lens.dev/posts/event/3.0.0.json",
  "lens": {
    "endsAt": "2028-10-01T01:00:00Z",
    "id": Any<String>,
    "links": [
      "https://example.com/tickets.html",
    ],
    "locale": "en",
    "location": "The Moon",
    "mainContentFocus": "EVENT",
    "position": "geo:40.6892,-74.0444",
    "startsAt": "2028-10-01T00:00:00Z",
    "title": "Where should we go next?",
  },
}
`;

exports[`Given the post metadata builders > when using the "event" builder > should return a valid EventMetadata for PhysicalAddress 1`] = `
{
  "$schema": "https://json-schemas.lens.dev/posts/event/3.0.0.json",
  "lens": {
    "address": {
      "country": "US",
      "locality": "New York",
      "postalCode": "10001",
      "region": "NY",
      "streetAddress": "1st Street",
    },
    "endsAt": "2028-10-01T01:00:00Z",
    "id": Any<String>,
    "locale": "en",
    "location": "The Moon",
    "mainContentFocus": "EVENT",
    "startsAt": "2028-10-01T00:00:00Z",
  },
}
`;

exports[`Given the post metadata builders > when using the "image" builder > should return a valid ImageMetadata 1`] = `
{
  "$schema": "https://json-schemas.lens.dev/posts/image/3.0.0.json",
  "lens": {
    "attachments": [
      {
        "altTag": "Me touching a tree",
        "item": "https://example.com/image-1.png",
        "license": "CC BY-NC",
        "type": "image/png",
      },
      {
        "altTag": "The tree touching me",
        "item": "https://example.com/image-2.png",
        "license": "CC BY-NC",
        "type": "image/png",
      },
    ],
    "id": Any<String>,
    "image": {
      "altTag": "Me touching grass",
      "item": "https://example.com/image.png",
      "license": "CCO",
      "type": "image/png",
    },
    "locale": "en",
    "mainContentFocus": "IMAGE",
    "title": "Touch grass",
  },
}
`;

exports[`Given the post metadata builders > when using the "link" builder > should return a valid LinkMetadata 1`] = `
{
  "$schema": "https://json-schemas.lens.dev/posts/link/3.0.0.json",
  "lens": {
    "content": "Check out this cool website!",
    "id": Any<String>,
    "locale": "en",
    "mainContentFocus": "LINK",
    "sharingLink": "https://example.com/embed.html",
  },
}
`;

exports[`Given the post metadata builders > when using the "liveStream" builder > should return a valid LiveStreamMetadata 1`] = `
{
  "$schema": "https://json-schemas.lens.dev/posts/livestream/3.0.0.json",
  "lens": {
    "id": Any<String>,
    "liveUrl": "https://example.com/live.html",
    "locale": "en",
    "mainContentFocus": "LIVESTREAM",
    "playbackUrl": "https://example.com/playback.html",
    "startsAt": "2028-10-01T00:00:00Z",
    "title": "Live stream #1",
  },
}
`;

exports[`Given the post metadata builders > when using the "liveStream" builder > when using the "mint" builder > should return a valid MintMetadata 1`] = `
{
  "$schema": "https://json-schemas.lens.dev/posts/mint/3.0.0.json",
  "lens": {
    "content": "Check out this NFT!",
    "id": Any<String>,
    "locale": "en",
    "mainContentFocus": "MINT",
    "mintLink": "https://opensea.io/assets/0x495f947276749ce646f68ac8c248420045cb7b5e/1234567890",
  },
}
`;

exports[`Given the post metadata builders > when using the "shortVideo" builder > should return a valid VideoMetadata 1`] = `
{
  "$schema": "https://json-schemas.lens.dev/posts/video/3.0.0.json",
  "lens": {
    "id": Any<String>,
    "locale": "en",
    "mainContentFocus": "SHORT_VIDEO",
    "title": "Great video!",
    "video": {
      "altTag": "The video of my life",
      "cover": "https://example.com/thumbnail.png",
      "duration": 123,
      "item": "https://example.com/video.mp4",
      "license": "CCO",
      "type": "video/mp4",
    },
  },
}
`;

exports[`Given the post metadata builders > when using the "space" builder > should return a valid SpaceMetadata 1`] = `
{
  "$schema": "https://json-schemas.lens.dev/posts/space/3.0.0.json",
  "lens": {
    "id": Any<String>,
    "link": "https://example.com/space.html",
    "locale": "en",
    "mainContentFocus": "SPACE",
    "startsAt": "2028-10-01T00:00:00Z",
    "title": "Space #1",
  },
}
`;

exports[`Given the post metadata builders > when using the "story" builder > should return a valid StoryMetadata 1`] = `
{
  "$schema": "https://json-schemas.lens.dev/posts/story/3.0.0.json",
  "lens": {
    "asset": {
      "altTag": "The story of my life",
      "cover": "https://example.com/thumbnail.png",
      "duration": 123,
      "item": "https://example.com/story.mp4",
      "type": "video/mp4",
    },
    "id": Any<String>,
    "locale": "en",
    "mainContentFocus": "STORY",
  },
}
`;

exports[`Given the post metadata builders > when using the "textOnly" builder > should return a valid TextOnlyMetadata 1`] = `
{
  "$schema": "https://json-schemas.lens.dev/posts/text-only/3.0.0.json",
  "lens": {
    "content": "GM!",
    "id": Any<String>,
    "locale": "en",
    "mainContentFocus": "TEXT_ONLY",
  },
}
`;

exports[`Given the post metadata builders > when using the "threeD" builder > should return a valid ThreeDMetadata 1`] = `
{
  "$schema": "https://json-schemas.lens.dev/posts/3d/3.0.0.json",
  "lens": {
    "assets": [
      {
        "format": "VRM",
        "playerUrl": "https://example.com/player.html",
        "uri": "https://example.com/model.zip",
        "zipPath": "foo/model.vrm",
      },
    ],
    "content": "Check out this 3D model!",
    "id": Any<String>,
    "locale": "en",
    "mainContentFocus": "3D",
    "tags": [
      "3d",
      "vrm",
    ],
  },
}
`;

exports[`Given the post metadata builders > when using the "transaction" builder > should return a valid TransactionMetadata 1`] = `
{
  "$schema": "https://json-schemas.lens.dev/posts/transaction/3.0.0.json",
  "lens": {
    "chainId": 1,
    "content": "Check out this transaction!",
    "id": Any<String>,
    "locale": "en",
    "mainContentFocus": "TRANSACTION",
    "txHash": "0x1234567890",
    "type": "ERC20",
  },
}
`;

exports[`Given the post metadata builders > when using the "video" builder > should return a valid VideoMetadata 1`] = `
{
  "$schema": "https://json-schemas.lens.dev/posts/video/3.0.0.json",
  "lens": {
    "attachments": [
      {
        "item": "https://example.com/soundtrack.mp3",
        "license": "CCO",
        "type": "audio/mpeg",
      },
    ],
    "id": Any<String>,
    "locale": "en",
    "mainContentFocus": "VIDEO",
    "title": "Great video!",
    "video": {
      "altTag": "The video of my life",
      "cover": "https://example.com/thumbnail.png",
      "duration": 123,
      "item": "https://example.com/video.mp4",
      "license": "CCO",
      "type": "video/mp4",
    },
  },
}
`;
